### PR TITLE
Add qa tier

### PR DESCRIPTION
**Pull request by ktp242**
This PR creates the configurations to add a QA tier to this Lambda function. For doing this, a function named CancelRequestResultConsumer-qa has been created.

https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/CancelRequestResultConsumer-qa?newFunction=true&tab=graph

The PR updates the configurations and env variables for connecting to the QA Lambda function. It also updates some out of date dependencies. Third, it adds the scripts for the CI/CD pipeline with Travis. Now, if a merge made to either development, qa, or master branch, and pushed, Travis will help to deploy that branch to a respective instance.

This branch has been manually deployed to CancelRequestResultConsumer-qa for testing.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,8 @@
 node_modules
 dist
 *.swp
-*.env
 vendor
 build
 .lambda
 .idea
+config/var_app
```
Comment (nonword) [inline .gitignore:8]: Nbd, but what's the need for the `config/var_app` ignore?

File: config/var_dev.env.example [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+OAUTH_CLIENT_ID=example_client_id
+OAUTH_CLIENT_SECRET=encrypted_client_secret
+OAUTH_TOKEN_URI=https://isso.nypl.org/oauth/token
+SCHEMA_BASE_URL=https://dev-platform.nypl.org/api/v0.1/current-schemas
+SLACK_CHANNEL=cancel-req-logging
+SLACK_LOGGING_LEVEL=info
+SLACK_TOKEN=encrypted_slack_token
+SLACK_USERNAME=CancelRequestResultConsumer-development
+SENDGRID_API_KEY=encrypted_sendgrid_api_key
+TIME_ZONE=America/New_York
```
Comment (nonword) [inline config/var_dev.env.example]: I don't personally think we need .example files after adding real, working config to source control. The real working config is the best example we can provide.

File: .travis.yml [hunk 3]
```diff
@@ -1,13 +1,38 @@
-language: php
+language:
+  - php
+  - node_js
 php:
-- 7.1
+  - 7.1
+node_js:
+  - 6.1
```
Comment (nonword) [inline .travis.yml]: We should probably specify node `6.10` for our tests - to match our `AWS_RUNTIME`

File: config/var_dev.env.example (hunk 2 above)
Comment (nodanaonlyzuul) [reply config/var_dev.env.example]: I agree with my friend @nonword.
If we commit the various stage/tier/environments' `.env` files than we can omit all the various `.env.sample` files.

File: scripts/travis-deploy.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (nodanaonlyzuul) [inline scripts/travis-deploy.sh:1]: I really like the inventiveness here...butttttt....

## Using .travis.yml to associate branches with environments

@danamansana and I have used a different approach for associating a branch to an environment. I think it's easy to read - and avoids the need for an intermediate script like this.
Look at [the bibservice's .travis.yml file](https://github.com/NYPL-discovery/bibservice/blob/master/.travis.yml) as an example. Let me know if you like it?

## Documenting which branch deploys to which environment

I've been putting [sections like this](https://github.com/NYPL/scsbuster/#git-workflow--deployment) in the README to document git workflow and how it relates to CI/CD. Would that be helpful in this repo?

File: .gitignore (hunk 1 above)
Comment (ktp242) [reply .gitignore:8]: `config/var_app` is only for the configurations to run the application locally. At this moment, we don't have any sensitive information in it, but maybe in the future, if the application updates, we might have. So I think ignoring it could help the developers run testing locally easier without encrypting the info.

File: config/var_dev.env.example (hunk 2 above)
Comment (ktp242) [reply config/var_dev.env.example]: Sure I can remove the examples.

File: .travis.yml (hunk 3 above)
Comment (ktp242) [reply .travis.yml]: Good idea.

File: scripts/travis-deploy.sh (hunk 4 above)
Comment (ktp242) [reply scripts/travis-deploy.sh:1]: So I got this alternative method from @nonword . I think the reason he suggested to use the script is for us to deploy to any environment locally without pushing it to remote. So I can deploy to QA for testing without really merging to QA branch.

I think I can add some documentations to .travis.yml so that it will be less discombobulating?

I can also add the branch => environment mapping to README.md

Thanks for the suggestions!

Comment (nonword) [reply scripts/travis-deploy.sh:1]: On the first point, @nodanaonlyzuul , I'm open to the idea that including a separate `deploy` entry for each of your environments is easier to read. For me the benefits of using a single, central `travis-deploy` script slightly outweighs that. These are those benefits:

1. to DRY,
2. to formalize the association between branch names and deployments (i.e. we have strong cross-repo conventions around what branches deploy where; I want to make that concrete and difficult to mis-type. [Scsbuster seems to take a similar approach fwiw](https://github.com/NYPL/scsbuster/blob/master/provisioning/travis_ci_and_cd/build_and_push_to_ecr.sh))
3. to support `deploy-` scripts that can be run as-is in both Travis and locally

That third point is maybe controversial and kind of depends on whether you view Travis as assistant or gate-keeper. Although I like to use Travis somewhat informally as gate-keeper, I've found it really useful to have `deploy-` scripts that can be run locally. The `--profile` is meaningless to Travis, so either you have to create a separate set of deploy scripts for use by Travis (which is error prone and a lot to read) or you have to do away with local deploying entirely. Although I think we should make it our practice to only ever deploy through Travis, I've found having the ability to quickly deploy to dev/qa in a pinch has been a big help in reducing the amount of time we have to stare at Travis waiting for it to spin up and finish a set of redundant tests just to confirm we got the security group right, for example. Maybe those rare occasions of lost time are worth the longer-term benefits of treating Travis as gate-keeper.

Comment (nodanaonlyzuul) [reply scripts/travis-deploy.sh:1]: Allo!

Well - I guess this boils down to preference. My preference is so skip the redirection of the intermediate script and have a longer .travis.yml file. (Both approaches' `package.json` remains the same length) That said it's easy enough to follow, and like I said in my original comment, I _do_ like its inventiveness and pragmatism.

About the final point - I guess I haven't had to deploy locally enough to feel bitten by having to add the `--profile` flag in the pinch where I'd have to deploy locally.

TL;DR: I'm fine with this approach.

Comment (ktp242) [reply scripts/travis-deploy.sh:1]: If we are ok with this approach,  I will make add some comments on `.travis.yml` for better documentations.

Comment (nodanaonlyzuul) [reply scripts/travis-deploy.sh:1]: Nah - I think it'd would be clear for folks to follow the breadcrumb trail to `deploy.sh`.

**Comment by ktp242** [comment]
I have updated the branch based on the comments!

File: README.md [hunk 5]
```diff
@@ -82,6 +80,14 @@ npm run test-event
 
 ## Deployment
```
Comment (nodanaonlyzuul) [inline README.md:82]: Sorry - I know it's a pain - but Can we add a line about how this is hooked up to CI/CD so merging to a branch automatically deploys?

Comment (ktp242) [reply README.md:82]: OK I have updated README.md

**Comment by nodanaonlyzuul** [comment]
I approve of this PR. Anyone have additional input? @nonword ?

**Comment by nonword** [comment]
Looks good to me! Thanks

**Review by nonword — APPROVED** [review]

**Review by nodanaonlyzuul — APPROVED** [review]
